contracts/LiquidityConversionRates.sol

```diff
@@ -0,0 +1,265 @@
+    uint public minBuyRateInPrecision;
+    uint public maxSellRateInPrecision;
+    uint public minSellRateInPrecision;
+
+    function LiquidityConversionRates(address _admin, ERC20 _token) public {
+        transferAdminQuickly(_admin);
+        token = _token;
+        setDecimals(token);
+        require(getDecimals(token) <= MAX_DECIMALS);
+    }
+
+    event ReserveAddressSet(address reserve);
+
+    function setReserveAddress(address reserve) public onlyAdmin {
+        reserveContract = reserve;
+        ReserveAddressSet(reserve);
+    }
+
+    event LiquidityParamsSet(
+        uint rInFp,
+        uint pMinInFp,
+        uint numFpBits,
+        uint maxCapBuyInFp,
+        uint maxEthCapSellInFp,
+        uint feeInBps,
+        uint formulaPrecision,
+        uint maxQtyInFp,
+        uint maxBuyRateInPrecision,
+        uint minBuyRateInPrecision,
+        uint maxSellRateInPrecision,
+        uint minSellRateInPrecision
+    );
+
+    function setLiquidityParams(
+        uint _rInFp,
+        uint _pMinInFp,
+        uint _numFpBits,
+        uint _maxCapBuyInWei,
+        uint _maxCapSellInWei,
+        uint _feeInBps,
+        uint _maxTokenToEthRateInPrecision,
+        uint _minTokenToEthRateInPrecision
+    ) public onlyAdmin {
+
+        require(_numFpBits < 256);
+        require(formulaPrecision <= MAX_QTY);
+        require(_feeInBps < 10000);
+        require(_minTokenToEthRateInPrecision < _maxTokenToEthRateInPrecision);
+
+        rInFp = _rInFp;
+        pMinInFp = _pMinInFp;
+        formulaPrecision = uint(1)<<_numFpBits;
+        maxQtyInFp = fromWeiToFp(MAX_QTY);
+        numFpBits = _numFpBits;
+        maxEthCapBuyInFp = fromWeiToFp(_maxCapBuyInWei);
+        maxEthCapSellInFp = fromWeiToFp(_maxCapSellInWei);
+        feeInBps = _feeInBps;
+        maxBuyRateInPrecision = PRECISION * PRECISION / _minTokenToEthRateInPrecision;
+        minBuyRateInPrecision = PRECISION * PRECISION / _maxTokenToEthRateInPrecision;
+        maxSellRateInPrecision = _maxTokenToEthRateInPrecision;
+        minSellRateInPrecision = _minTokenToEthRateInPrecision;
+
+        LiquidityParamsSet(
+            rInFp,
+            pMinInFp,
+            numFpBits,
+            maxEthCapBuyInFp,
+            maxEthCapSellInFp,
+            feeInBps,
+            formulaPrecision,
+            maxQtyInFp,
+            maxBuyRateInPrecision,
+            minBuyRateInPrecision,
+            maxSellRateInPrecision,
+            minSellRateInPrecision
+        );
+    }
+
+    function recordImbalance(
+        ERC20 conversionToken,
+        int buyAmountInTwei,
+        uint rateUpdateBlock,
+        uint currentBlock
+    )
+        public
+    {
+        conversionToken;
+        rateUpdateBlock;
+        currentBlock;
+
+        require(msg.sender == reserveContract);
+        if (buyAmountInTwei > 0) {
+            // Buy case
+            collectedFeesInTwei += calcCollectedFee(abs(buyAmountInTwei));
```

Comment:
it is intentionally in past notation, since recordimbalance is called at the end of the trade process, after the fees where collected in some way.
also - we don't really know if the fees have been collected by the reserve manager or not (i,e - if tokens where withdrawn for covering knc fees).
so leaving name as is for now.